ctru-rs/src/services/ir_user.rs

```diff
@@ -0,0 +1,491 @@
+            machine_id: shared_mem[11],
+            unknown_field_1: shared_mem[12],
+            network_id: shared_mem[13],
+            unknown_field_2: shared_mem[14],
+            unknown_field_3: shared_mem[15],
+        }
+    }
+
+    /// Read and parse the current packets received from the IR device.
+    pub fn get_packets(&self) -> Vec<IrUserPacket> {
+        let shared_mem_guard = IR_USER_STATE.lock().unwrap();
+        let user_state = shared_mem_guard.as_ref().unwrap();
+        let shared_mem = user_state.shared_memory;
+
+        // Find where the packets are, and how many
+        let start_index = u32::from_ne_bytes([
+            shared_mem[0x10],
+            shared_mem[0x11],
+            shared_mem[0x12],
+            shared_mem[0x13],
+        ]);
+        let valid_packet_count = u32::from_ne_bytes([
+            shared_mem[0x18],
+            shared_mem[0x19],
+            shared_mem[0x1a],
+            shared_mem[0x1b],
+        ]);
+
+        // Parse the packets
+        (0..valid_packet_count as usize)
+            .map(|i| {
+                // Get the packet info
+                let packet_index = (i + start_index as usize) % user_state.recv_packet_count;
+                let packet_info_offset =
+                    SHARED_MEM_RECV_BUFFER_OFFSET + (packet_index * PACKET_INFO_SIZE);
+                let packet_info =
+                    &shared_mem[packet_info_offset..packet_info_offset + PACKET_INFO_SIZE];
+
+                let offset_to_data_buffer = u32::from_ne_bytes([
+                    packet_info[0],
+                    packet_info[1],
+                    packet_info[2],
+                    packet_info[3],
+                ]) as usize;
+                let data_length = u32::from_ne_bytes([
+                    packet_info[4],
+                    packet_info[5],
+                    packet_info[6],
+                    packet_info[7],
+                ]) as usize;
+
+                // Find the packet data. The packet data may wrap around the buffer end, so
+                // `packet_data` is a function from packet byte offset to value.
+                let packet_info_section_size = user_state.recv_packet_count * PACKET_INFO_SIZE;
+                let header_size = SHARED_MEM_RECV_BUFFER_OFFSET + packet_info_section_size;
+                let data_buffer_size = user_state.recv_buffer_size - packet_info_section_size;
+                let packet_data = |idx| -> u8 {
+                    let data_buffer_offset = offset_to_data_buffer + idx;
+                    shared_mem[header_size + data_buffer_offset % data_buffer_size]
+                };
+
+                // Find out how long the payload is (payload length is variable-length encoded)
+                let (payload_length, payload_offset) = if packet_data(2) & 0x40 != 0 {
+                    // Big payload
+                    (
+                        ((packet_data(2) as usize & 0x3F) << 8) + packet_data(3) as usize,
```

Comment:
I believe this should be `((packet_data(2) as usize & 0x3F) << 6) + packet_data(3) as usize`

The first bit contains upper 6 bits rather than half of the payload size.